pkg/hub/controllers/placemanifestworkcontroller/placemanifestwork_deploy_reconcile.go

```diff
@@ -2,21 +2,170 @@ package placemanifestworkcontroller
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
 	clusterlister "open-cluster-management.io/api/client/cluster/listers/cluster/v1beta1"
-	workclientset "open-cluster-management.io/api/client/work/clientset/versioned"
+	//workclientset "open-cluster-management.io/api/client/work/clientset/versioned"
 	worklisterv1 "open-cluster-management.io/api/client/work/listers/work/v1"
+	"open-cluster-management.io/api/utils/work/v1/workapplier"
+	workv1 "open-cluster-management.io/api/work/v1"
 	workapiv1alpha1 "open-cluster-management.io/api/work/v1alpha1"
+	"open-cluster-management.io/work/pkg/helper"
 )
 
 // deployReconciler is to manage ManifestWork based on the placement.
 type deployReconciler struct {
-	workClient          workclientset.Interface
+	//workClient          workclientset.Interface
+	workApplier         *workapplier.WorkApplier
 	manifestWorkLister  worklisterv1.ManifestWorkLister
 	placeDecisionLister clusterlister.PlacementDecisionLister
 	placementLister     clusterlister.PlacementLister
 }
 
 func (d *deployReconciler) reconcile(ctx context.Context, pw *workapiv1alpha1.PlaceManifestWork) (*workapiv1alpha1.PlaceManifestWork, reconcileState, error) {
 	// TODO the main manifestwork create/update/delete logic, we may also need to manage the status condition for whether manifestwork is created here
-	return pw, reconcileContinue, nil
+	klog.Info("reconciling placeManifestWork ", pw.Name)
+
+	placement, err := d.placementLister.Placements(pw.Namespace).Get(pw.Spec.PlacementRef.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			apimeta.SetStatusCondition(&pw.Status.Conditions, GetPlacementDecisionVerified(workapiv1alpha1.PlacementDecisionNotFound, ""))
+		}
+		return pw, reconcileContinue, fmt.Errorf("Failed get placement %w", err)
+	}
+
+	//klog.Info("my reconciling placement ", placement)
+
+	req, err := labels.NewRequirement(PlaceManifestWorkControllerNameLabelKey, selection.In, []string{pw.Name})
+	if err != nil {
+		return pw, reconcileContinue, err
+	}
+
+	selector := labels.NewSelector().Add(*req)
+	manifestWorks, err := d.manifestWorkLister.List(selector)
+	if err != nil {
+		return pw, reconcileContinue, err
+	}
+
+	errs := []error{}
+	existingClusters := sets.NewString()
+	for _, mw := range manifestWorks {
+		existingClusters.Insert(mw.Namespace)
+	}
+
+	addedClusters, deletedClusters, err := helper.GetClusters(d.placeDecisionLister, placement, existingClusters)
+	if err != nil {
+		return pw, reconcileContinue, utilerrors.NewAggregate(errs)
+	}
+	//klog.Info("Selected cluters count ", len(added), len(deleted))
+
+	// Delete manifestWork for deleted clusters
+	for cls := range deletedClusters {
+		err = d.workApplier.Delete(ctx, cls, pw.Name)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	// Create manifestWork for added clusters
+	for cls := range addedClusters {
```

Comment:
how about apply on all existing and added clusters then you would not need line 91~103

Reply:
okay, its just the existingClusters also contain the deleted clusters, that is reason for deleting first. Will make the loop on the existingCluster and then apply with the pmw.spec.mw assigned to the mw to be sure that if the mw.spec had changed we still enforce the pwm.spec.mw